### PR TITLE
Add JobIncrementer to CTR

### LINKED ISSUE #9: Add a configurable `JobParametersIncrementer` to allow users to run a task flow multiple times easily
This comes from the issue here: https://github.com/spring-cloud/spring-cloud-dataflow/issues/1489.  While Spring Batch is used to orchestrate a composed task, a user may not be interested in the restartability/etc that Spring Batch provides.  Providing a `JobParametersIncrementer` that is configurable here would allow users to either get that functionality or easily launch a flow repeatedly.

### PR DESCRIPTION
**Pull request by cppwfs**
resolves #9

### CODE REVIEW

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedTaskJobIncrementer.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.batch.core.JobParametersBuilder;
+import org.springframework.batch.core.JobParametersIncrementer;
+
+/**
+ * Supports ability for a ComposedTaskRunner to create a new Job Instance on
+ * subsequent runs.
+ *
+ * @author Glenn Renfro
+ */
+public class ComposedTaskJobIncrementer implements JobParametersIncrementer {
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedTaskJobIncrementer.java]: Why a custom `JobParametersIncrementer` and not the OOTB `RunIdIncrementer`: https://docs.spring.io/spring-batch/apidocs/org/springframework/batch/core/launch/support/RunIdIncrementer.html

File: spring-cloud-starter-task-composedtaskrunner/README.adoc [hunk 2]
```diff
@@ -152,7 +151,9 @@ $$split-thread-queue-capacity$$:: $$Capacity for Split's  BlockingQueue.
 $$split-thread-wait-for-tasks-to-complete-on-shutdown$$:: $$Whether to wait for scheduled tasks to complete on shutdown, not
  interrupting running tasks and executing all tasks in the queue.
  Default is false;$$ *($$Boolean$$, default: `$$false$$`)*
-
+$$unique-instance-enabled$$:: $$Allows a single ComposedTaskRunner instance to be re-executed. Default
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/README.adoc]: This name is inaccurate.  It should be something like `increment-instance-enabled` IMHO.  Also, the text should state that it allows a single `ComposedTaskRunner` instance to be re-executed _without changing the parameters_.  You can always re-execute a `ComposedTaskRunner` if you change the parameters.  This allows you to not need to do so.

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerJobFactory.java [hunk 3]
```diff
@@ -89,13 +94,16 @@ public Job getObject() throws Exception {
...
-		return this.jobBuilderFactory
+		FlowJobBuilder builder = this.jobBuilderFactory
 				.get(this.taskNameResolver.getTaskName())
 				.start(this.flowBuilder
 						.start(createFlow())
 						.end())
-				.end()
-				.build();
+				.end();
+		if(uniqueInstanceEnabled) {
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerJobFactory.java]: `this.unieuqInstanceEnabled`

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java [hunk 4]
```diff
@@ -111,6 +111,13 @@
 	 */
 	private boolean splitThreadWaitForTasksToCompleteOnShutdown;
 
+	/**
+	 * Allows a single ComposedTaskRunner instance to be re-executed. Default
+	 * is false which means a ComposedTaskRunner instance can only be executed once, if
+	 * true it can be re-executed.
+	 */
+	private boolean uniqueInstanceEnabled = false;
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java]: See previous note about the name of this field

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java [hunk 5]
```diff
@@ -215,4 +222,12 @@ public boolean isSplitThreadWaitForTasksToCompleteOnShutdown() {
 	public void setSplitThreadWaitForTasksToCompleteOnShutdown(boolean splitThreadWaitForTasksToCompleteOnShutdown) {
 		this.splitThreadWaitForTasksToCompleteOnShutdown = splitThreadWaitForTasksToCompleteOnShutdown;
 	}
+
+	public boolean isUniqueInstanceEnabled() {
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java]: See previous about the name of this field.

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java [hunk 6]
```diff
@@ -215,4 +222,12 @@ public boolean isSplitThreadWaitForTasksToCompleteOnShutdown() {
...
+		return uniqueInstanceEnabled;
+	}
+
+	public void setUniqueInstanceEnabled(boolean uniqueInstanceEnabled) {
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java]: See previous about the name of this field.

File: spring-cloud-starter-task-composedtaskrunner/src/test/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedTaskRunnerConfigurationJobIncrementerTests.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	@Autowired
+	private TaskOperations taskOperations;
+
+	@Test
+	@DirtiesContext
+	public void testComposedConfigurationWithJobIncrementer() throws Exception {
+		JobExecution jobExecution = this.jobRepository.createJobExecution(
+				"ComposedTest", new JobParameters());
+		job.execute(jobExecution);
+		Assert.notNull(job.getJobParametersIncrementer(), "JobParametersIncrementer must not be null.");
```
Comment (mminella) [inline spring-cloud-starter-task-composedtaskrunner/src/test/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedTaskRunnerConfigurationJobIncrementerTests.java]: You don't need to execute the job for this assertion.

**Review by mminella — CHANGES_REQUESTED** [review]

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerJobFactory.java (hunk 3 above)
Comment (cppwfs) [reply spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/ComposedRunnerJobFactory.java]: fixed

File: spring-cloud-starter-task-composedtaskrunner/README.adoc (hunk 2 above)
Comment (cppwfs) [reply spring-cloud-starter-task-composedtaskrunner/README.adoc]: Fixed

File: spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java (hunk 6 above)
Comment (cppwfs) [reply spring-cloud-starter-task-composedtaskrunner/src/main/java/org/springframework/cloud/task/app/composedtaskrunner/properties/ComposedTaskProperties.java]: fixed

**Comment by mminella** [comment]
LGTM.  Squashed and merged as c3770c4f367c6fd0bf1eef93005687fcebbbb0ca

**Pull request closed without merging**